flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java

```diff
@@ -596,6 +596,36 @@ public void testSharedIncrementalStateDeRegistration() throws Exception {
         }
     }
 
+    @Test
+    public void testDisposeCleanRocksDBStateDataTransferThread() throws Exception {
+        setupRocksKeyedStateBackend();
+        try {
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot =
+                    keyedStateBackend.snapshot(
+                            0L,
+                            0L,
+                            testStreamFactory,
+                            CheckpointOptions.forCheckpointWithDefaultLocation());
+            Thread asyncSnapshotThread = new Thread(snapshot);
+            asyncSnapshotThread.start();
+            waiter.await(); // wait for snapshot to run
+            waiter.reset();
+            runStateUpdates();
+            blocker.trigger(); // allow checkpointing to start writing
+            waiter.await(); // wait for snapshot stream writing to run
+            SnapshotResult<KeyedStateHandle> snapshotResult = snapshot.get();
+        } finally {
+            IOUtils.closeQuietly(keyedStateBackend);
+            keyedStateBackend.dispose();
+        }
+        for (Thread t : Thread.getAllStackTraces().keySet()) {
+            if (t.getName().contains("Flink-RocksDBStateDataTransfer")
+                    && t.getState() != Thread.State.TERMINATED) {
```

Comment:
This check depends on the call to `ExecutorService.shutdownNow`. Its javadoc says:
> This method does not wait for actively executing tasks to terminate. Use awaitTermination to do that.

So I'm afraid the check may fail from time to time.

Another issue IMO is that relying on the thread name is a bit fragile, as it's not explicit contract.

Maybe we can create a new `RocksDBStateUploader` that overrides `close()` and pass it the strategy from builder?
So that later we can check whether its `close()` method was called or not.

Reply:
Thank you very much for your review  @rkhachatryan .I totally agree with you option, I will fix it as soon as possible